source/common/grpc/transcoding_filter.cc

```diff
@@ -0,0 +1,396 @@
+#include "grpc_transcoding/response_to_json_translator.h"
+
+using google::grpc::transcoding::JsonRequestTranslator;
+using google::grpc::transcoding::RequestInfo;
+using google::grpc::transcoding::ResponseToJsonTranslator;
+using google::grpc::transcoding::Transcoder;
+using google::grpc::transcoding::TranscoderInputStream;
+using google::protobuf::DescriptorPool;
+using google::protobuf::FileDescriptor;
+using google::protobuf::FileDescriptorSet;
+using google::protobuf::io::ZeroCopyInputStream;
+using google::protobuf::util::error::Code;
+using google::protobuf::util::Status;
+
+namespace Envoy {
+namespace Grpc {
+
+namespace {
+
+const std::string kTypeUrlPrefix{"type.googleapis.com"};
+
+// Transcoder implementation based on JsonRequestTranslator &
+// ResponseToJsonTranslator
+class TranscoderImpl : public Transcoder {
+public:
+  // request_translator - a JsonRequestTranslator that does the request
+  //                      translation
+  // response_translator - a ResponseToJsonTranslator that does the response
+  //                       translation
+  TranscoderImpl(std::unique_ptr<JsonRequestTranslator> request_translator,
+                 std::unique_ptr<ResponseToJsonTranslator> response_translator)
+      : request_translator_(std::move(request_translator)),
+        response_translator_(std::move(response_translator)),
+        request_stream_(request_translator_->Output().CreateInputStream()),
+        response_stream_(response_translator_->CreateInputStream()) {}
+
+  // Transcoder implementation
+  ::google::grpc::transcoding::TranscoderInputStream* RequestOutput() {
+    return request_stream_.get();
+  }
+  Status RequestStatus() { return request_translator_->Output().Status(); }
+
+  ZeroCopyInputStream* ResponseOutput() { return response_stream_.get(); }
+  Status ResponseStatus() { return response_translator_->Status(); }
+
+private:
+  std::unique_ptr<JsonRequestTranslator> request_translator_;
+  std::unique_ptr<ResponseToJsonTranslator> response_translator_;
+  std::unique_ptr<::google::grpc::transcoding::TranscoderInputStream> request_stream_;
+  std::unique_ptr<::google::grpc::transcoding::TranscoderInputStream> response_stream_;
+};
+
+} // namespace
+
+TranscodingConfig::TranscodingConfig(const Json::Object& config) {
+  std::string proto_descriptor_file = config.getString("proto_descriptor");
+  FileDescriptorSet descriptor_set;
+  if (!descriptor_set.ParseFromString(Filesystem::fileReadToEnd(proto_descriptor_file))) {
+    throw EnvoyException("Unable to parse proto descriptor");
+  }
+
+  for (const auto& file : descriptor_set.file()) {
+    if (descriptor_pool_.BuildFile(file) == nullptr) {
+      throw EnvoyException("Unable to parse proto descriptor");
+    }
+  }
+
+  google::grpc::transcoding::PathMatcherBuilder<const google::protobuf::MethodDescriptor*> pmb;
+
+  for (const auto& service_name : config.getStringArray("services")) {
+    auto service = descriptor_pool_.FindServiceByName(service_name);
```

Comment:
I tried to do this before and didn't do that since:
- we use service config in Cloud Endpoints, which is compiled from descriptor, so envoy is currently only one customer use descriptor set directly.
- path_matcher library is separated from transcoding, doesn't have protobuf dependency.

I added a TODO here to consider factoring this out, along with Endpoints.